src/transformStateWithClones.js

```diff
@@ -7,7 +7,39 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateHistory = [];
+  let stateCopy = { ...state };
+
+  for (const action of actions) {
+    stateCopy = { ...stateCopy };
+
+    switch (action.type) {
+      case 'addProperties':
+        for (const key in action.extraData) {
+          stateCopy[key] = action.extraData[key];
+        }
+        break;
+
+      case 'removeProperties':
+        for (const key of action.keysToRemove) {
+          delete stateCopy[key];
+        }
+        break;
+
+      case 'clear':
+        for (const key in stateCopy) {
+          delete stateCopy[key];
+        }
```

Comment:
[Checklist](https://github.com/mate-academy/js_stateful-clones/blob/master/checklist.md) p. 4, you can create function for removing properties by key like 

> function deleteByKey(dynamicArray, keysArray = dynamicArray) {
>     for (const key in keysArray) {
>       delete dynamicArray[key];
>     }
>   }

and here call `deleteByKey(stateCopy);`
in case removeProperties call `deleteByKey(stateCopy, action.keysToRemove);`